compiler/optimizer/OMRValuePropagation.cpp

```diff
@@ -6393,6 +6405,23 @@ void OMR::ValuePropagation::buildBoundCheckComparisonNodes(BlockVersionInfo *blo
                traceMsg(comp(), "Second test - Creating %p (%s)\n", nextComparisonNode, nextComparisonNode->getOpCode().getName());
 
             temp.add(nextComparisonNode);
+
+            if (arrayIndex->_baseNode && arrayIndex->_instanceOfClass && !comp()->compileRelocatableCode() && 
```

Comment:
The !compileRelocatableCode seems like a bit of a chicken out - if we can relocate the class we can do the opt. @dsouzai is there an easy way with the new infrastructure to allow this to work under AOT?

Reply:
```
// InstanceOf check for the object were we load the array index from
TR::Node *objectRefChild = arrayIndex->_baseNode->getFirstChild();
```
In order for this tree to have bene created, I'm guessing we must have made a frontend query to know the instanceof relationship. If that's true, then the validation would've already been created for AOT. If using non-SVM AOT, then the frontend query would've returned NULL, in which case the compiler would not know any instanceof information. 

As such, I'm not sure if the AOT guard is needed. That said, if there really is a concern, you can always make a call to the frontend to do the instanceof check, and if it passes, then go down the code path.

Reply:
Any reply to this @klangman ?

Reply:
The instanceOf-check generated below (one that existed before I changed anything here) has this AOT check as well. I can't say I fully understand why, it seems to me that the instanceOf-checks are needed for correctness. I would be happy to disable this for AOT, but I guess we might need someone from AOT land to help understand why we can get away without the instanceOf-check in the case of AOT.

Reply:
Hmm - if the check is for correctness I can't see why they can be omitted for AOT without risking violating the correctness. Would you agree @dsouzai ?

Reply:
I had a phone call with @klangman to understand the full picture here, and we both agreed that the old code that didn't emit the instanceof checks under AOT was wrong. That said, I know why the code was guarded:

```
TR::Node *duplicateClassPtr = TR::Node::createWithSymRef(objectRefChild, TR::loadaddr, 0, comp()->getSymRefTab()->findOrCreateClassSymbol(objectRef->getSymbolReference()->getOwningMethodSymbol(comp()), -1, arrayLength->_instanceOfClass, false));
```
passes in a cpIndex of -1, which non SVM AOT can't handle. My guess is that whoever implemented this saw weird crashes under AOT and did the usual "I'm just gonna disable this under AOT".

Technically there should be no problem in SVM AOT; however, to keep the scope of this PR from getting bigger and bigger, both Kevin and I agreed that the best course for now is to prevent versioning if AOT && instanceOf checks were added.

In a future PR, we can enable versioning in AOT when the SVM is enabled.